compiler/rewrite-rules/early_opt.rules

```diff
@@ -0,0 +1,79 @@
+;;; Function specialization rules
+
+;; Using a simple pattern matching in combination with constant folding,
+;; replace functions with their specialized forms.
+;; The specialized forms have better typing.
+
+(op_func_call {"microtime"})            =>    (op_func_call {"_microtime_string"})
+(op_func_call {"microtime"} (op_false)) => (op_func_call {"_microtime_string"})
+(op_func_call {"microtime"} (op_true))  => (op_func_call {"_microtime_float"})
+
+(op_func_call {"hrtime"})            => (op_func_call {"_hrtime_array"})
+(op_func_call {"hrtime"} (op_false)) => (op_func_call {"_hrtime_array"})
+(op_func_call {"hrtime"} (op_true))  => (op_func_call {"_hrtime_int"})
+
+;;; String conversion optimization rules
+
+(op_string_build x) => (op_conv_string x)
+
+(op_concat (op_string {""}) x) => (op_conv_string x)
+(op_concat x (op_string {""})) => (op_conv_string x)
+
+;; eager string conversion may result in redundant string allocation
+;; let concat overloadings do their magic: maybe there is an overloading for
+;; a given type that can be appended to the string directly
+(op_concat (op_conv_string x) y) => (op_concat x y)
+(op_concat x (op_conv_string y)) => (op_concat x y)
+
+;;; Constant folding rules
+
+(op_func_call {"ord"} arg:(op_string))
+  where { arg->str_val.size() == 1 }
+  => (op_int_const { std::to_string(static_cast<unsigned char>(arg->str_val[0])) })
+
+(op_func_call {"chr"} arg:(op_int_const))
+  => (op_string { std::string(1, std::stol(arg->str_val)) })
+
+(op_func_call {"strlen"} arg:(op_string))
+  => (op_int_const { std::to_string(arg->str_val.size()) })
+
+(op_func_call {"count"} arg:(op_array ...))
+  => (op_int_const { std::to_string(arg->args().size()) })
+
+;;; Substr (and alike) optimization rules
```

Comment:
Can you please add small examples (here and above) about what's going on? What PHP code snippets are expected, and what they are transformed into